partner_compassion/models/survey.py

```diff
@@ -0,0 +1,114 @@
+##############################################################################
+#
+#    Copyright (C) 2014 Compassion CH (http://www.compassion.ch)
+#    Releasing children from poverty in Jesus' name
+#    @author: Quentin Gigon <gigon.quentin@gmail.com>
+#
+#    The licence is in the file __manifest__.py
+#
+##############################################################################
+
+from odoo import models, fields, api
+
+
+class SurveyQuestion(models.Model):
+    _inherit = "survey.question"
+
+    # when used this field tells the validation process that the maximum number of option answer should be validated
+    max_checked_option = fields.Integer(
+        string='if used_max_check_options. maximum number of options allowed for the user')
+
+    @api.multi
+    def validate_multiple_choice(self, post, answer_tag):
+        # call super() function to avoid missing important behaviour
+        errors = super().validate_multiple_choice(post=post, answer_tag=answer_tag)
+        # add check for maximum number of answered option.
+        if self.max_checked_option:
+            # get answer candidate based on post dictionary
+            answer_candidates = {k: v for k, v in post.items() if k.startswith(answer_tag)}
+            if len(answer_candidates) > self.max_checked_option:
+                errors.update({answer_tag: self.constr_error_msg})
+
+        return errors
+
+
+class SurveyLabel(models.Model):
+    _inherit = "survey.label"
+
+    order_in_question = fields.Integer(
+        'Order of the question. Used in multiple options question when order is important. should starts at 0',
+    )
+
+
+class SurveyUserInput(models.Model):
+    _inherit = "survey.user_input"
+
+    def write(self, vals):
+        out = super(SurveyUserInput, self).write(vals)
+
+        # compute segment affinity if needed
+        state = 'state' in vals and vals['state'] == 'done'
+        _id = self.survey_id == self.env.ref('partner_compassion.partner_segmentation_survey')
+        if state and _id:
+            ans = self._get_answer_as_array()
+            self.env['res.partner.segment.affinity'].segment_affinity_engine(ans, self.partner_id)
```

Comment:
You can directly pass the id as you don't need the whole record.
```suggestion
            self.env['res.partner.segment.affinity'].segment_affinity_engine(ans, self.partner_id.id)
```